ports/stm32: stm32wl5x SUBGHZ radio modem.

projectgus opened the pull request:
Adds support for the STM32WL5's LoRa SUBGHZ modem & its associated SPI interface. Plus some small cleanup commits for DMA to work on STM32WL5x.

The C part of this interface is aiming for the minimum amount of C code to support a LoRa modem driver written in Python. This PR is a dependency for a micropython-lib update adding the LoRa driver (see https://github.com/micropython/micropython-lib/pull/572).

Additional commit adds a new build flags `MICROPY_PY_STM_CONST`, which defaults to 1 if `MICROPY_PY_STM` is set but can be set to 0 to generate the `stm` module without the named register constants, which take up quite a bit of code size.

## TODOs

* [x] Documentation for the new modem class
* [x] Determine if freezing the Python driver is the correct approach

## Follow-up for future PRs

* Freeze the Python driver into the stm32wl5 build by default (see 288a9cc829)

This work was funded by GitHub sponsors.

projectgus committed:
stm32/spi: Add STM32WL5 SUBGHZ SPI peripheral.

This is a "normal" SPI peripheral with no external pins, to avoid having to
grow spi_obj[] for just this one board map it as SPI ID 3 (board has SPI
IDs 1,2 already).

Signed-off-by: Angus Gratton <angus@redyak.com.au>

projectgus committed:
stm32/dma: Fix DMA completion on WL55 boards.

No IRQHandlers were compiled in for this board.

Includes small consolidation of the same DMAMUX_ENABLE line for STM32G4,
STM32WB, STM32WL.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

projectgus committed:
stm32/subghz: Add STM32WL55 subghz radio interface to stm module.

This is the minimum C interface to allow a modem driver to be built in
Python.  Interface is simple, with the intention that the micropython-lib
driver is the main (only) consumer of it.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

projectgus commented on ports/stm32/boards/NUCLEO_WL55/manifest.py:
```diff
@@ -0,0 +1,2 @@
+include("$(MPY_DIR)/extmod/uasyncio/manifest.py")
+freeze("$(MPY_LIB_DIR)/micropython/lora", ("async_modem.py", "stm32wl5.py"))
```
Pros/cons for including this modem by default:

* Pro: Likely anyone using this board will be using a LoRa modem, as why else would you have it?
* Pro: stm32wl5.py source code is too big to fit into the small RAM on this device, needs to always be loaded as bytecode. Freezing it in makes it easier for users to get it working in the device.
* Con: Makes default image larger (+15KB, text size 219504->235104, of which about 4.5KB is LoRa modem and the rest if asyncio). 
* Con: Assumes user will want this modem and not an alternative modem implementation or version.

Can easily drop this commit if it's too much overhead.

A midway compromise could be to freeze in the synchronous modem driver stm32wl5.py (approx 3.5KB) but not the async support.

dpgeorge replied:
I think we should freeze it in by default, because of your first point (why else would you use this board?).  And asyncio is so useful that we should also include that.

There are things we can do to reduce firmware size in other ways by removing other features that are less used (but that can be done later if needed).

projectgus commented:
@dpgeorge @jimmo PTAL. Have moved out the micropython-lib & manifest change for a future PR (after the micropython-lib PRs have landed), and also added some docs for this kinda weird hardware interface.

projectgus commented:
Rebased to resolve merge conflict.

dpgeorge commented on ports/stm32/stm32_it.c:
```diff
@@ -1033,6 +1034,14 @@ void LPUART2_IRQHandler(void) {
+    handle_subghz_radio_irq();
+    IRQ_EXIT(SUBGHZ_Radio_IRQn);
+}
+#endif
```
This can go in `subghz_radio.c`, and just wrap the handler directly there.

dpgeorge commented on ports/stm32/subghz_radio.c:
```diff
@@ -0,0 +1,193 @@
+    { MP_ROM_QSTR(MP_QSTR_cs), MP_ROM_PTR(&machine_subghz_radio_cs_obj) },
+    { MP_ROM_QSTR(MP_QSTR_irq), MP_ROM_PTR(&machine_subghz_radio_irq_obj) },
+    { MP_ROM_QSTR(MP_QSTR_irq_en), MP_ROM_PTR(&machine_subghz_radio_irq_en_obj) },
+    { MP_ROM_QSTR(MP_QSTR_is_busy), MP_ROM_PTR(&machine_subghz_radio_is_busy_obj) },
```
Make these into 3 functions (subghz_cs, subghz_irq, subghz_is_busy) in the `stm` module.

projectgus replied:
Done. Needed a fourth function as well, to enable/disable the HSE32 oscillator.

dpgeorge commented on ports/stm32/subghz_radio.c:
```diff
@@ -0,0 +1,193 @@
+
+    if (!self->spi) {
+        /* Instantiate the SUBGHZ SPI object and keep a pointer to it here.
+           The new SPI object won't point to the heap, it points to the static array of SPI instances. */
```
use C++ style comments

dpgeorge commented on ports/stm32/subghz_radio.c:
```diff
@@ -0,0 +1,193 @@
+
+       Rather than writing to SUBGHZ SPI in Interrupt Context to clear the interrupt, disable the IRQ
+       and rely on the radio driver to call radio.irq_en(True) to re-enable when needed.
+    */
```
C++ style comments

projectgus replied:
I finally have these highlighted as errors in emacs!

dpgeorge commented on ports/stm32/main.c:
```diff
@@ -528,6 +529,10 @@ void stm32_main(uint32_t reset_mode) {
     #endif
 
+    #if MICROPY_HW_ENABLE_SUBGHZ_RADIO
+    machine_subghz_radio_init();
```
This would be called `stm_subghz_radio_init()`.

projectgus replied:
I went with `subghz_init()`, in keeping with the naming used for the `rfcore` subsystem and the suggested prefix for the Python functions.

projectgus commented:
@dpgeorge @jimmo  Reworked as per review coments, PTAL when you have a moment.

Code size impact (without any LoRa code from micropython-lib).

Before (master):

```
   text    data     bss     dec     hex filename
 215588      80   19964  235632   39870 build-NUCLEO_WL55/firmware.elf
```

After (this branch):

```
   text    data     bss     dec     hex filename
 225648      80   20124  245852   3c05c build-NUCLEO_WL55/firmware.elf
```

+10,220 bytes, it's a bit! 

However, if we take the master branch and just enable  `MICROPY_PY_STM`:

```
   text    data     bss     dec     hex filename
 224480      80   19964  244524   3bb2c build-NUCLEO_WL55/firmware.elf
```

+ 8,892 bytes, suspect much of which might be the `stmconst` named register access.

What do you think about adding a build flag `MICROPY_PY_STM_CONST` or similar, to allow adding the `stm` module without the register generation?

dpgeorge commented:
> 8,892 bytes, suspect much of which might be the stmconst named register access

Yes that would be exactly what it is.

> What do you think about adding a build flag MICROPY_PY_STM_CONST or similar, to allow adding the stm module without the register generation?

Yes that's a good idea, because the WL doesn't have much flash and we need the `stm` module for the subgz functions.

projectgus committed:
stm32/modstm: Add MICROPY_PY_STM_CONST flag, clear it for STM32WL5.

MICROPY_PY_STM_CONST defaults to 1 if MICROPY_PY_STM is set.  Overriding to
0 disables the named register peripheral constants being including in the
stm32 module.

This saves about 7.5KB of code size for the STM32WL55, which is significant
as this SoC doesn't have a lot of flash.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

projectgus commented:
Done!

```
   text    data     bss     dec     hex filename
 216860      80   20124  237064   39e08 build-NUCLEO_WL55/firmware.elf
```

This PR adds only 1432 bytes now (for subghz functions and the minimal stm module.)

dpgeorge commented on docs/library/stm.rst:
```diff
@@ -102,3 +102,52 @@ the second CPU, the RF core.
+   Sets the internal SPI CS pin attached to the radio peripheral. The
+   ``deassert`` argument has the same behaviour as calling :ref:`machine.Pin()
+   <machine.Pin>` to set the output: a truthy value means de-asserted, falsey means
+   asserted.
```
I think this needs a little more explanation, like "a truthy value means "CS pin high" which deasserts the line.  a falsey value means "CS pin low" which asserts it."

dpgeorge commented on ports/stm32/boards/NUCLEO_WL55/mpconfigboard.h:
```diff
@@ -48,6 +49,10 @@
 
+// SUBGHZSPI Internal radio SPI BUS
+#define MICROPY_HW_SUBGHZSPI_NAME "SUBGHZ"
+#define MICROPY_HW_SUBGHZSPI_ID 3
```
please indent to match other lines in this file

dpgeorge commented on ports/stm32/main.c:
```diff
@@ -528,6 +529,10 @@ void stm32_main(uint32_t reset_mode) {
 
+    #if defined(STM32WL)
+    subghz_init();
+    #endif
```
IMO this should go next to rfcore_init().  

Then there should be a separate subghz_deinit() which just clears the callback handler, before gc_sweep_all() is run.  Otherwise a hard interrupt that comes in after gc_sweep_all() is called, will call bad data.

projectgus replied:
> IMO this should go next to rfcore_init().
>
> Then there should be a separate subghz_deinit() which just clears the callback handler, before gc_sweep_all() is run. Otherwise a hard interrupt that comes in after gc_sweep_all() is called, will call bad data.

Oh, I missed that root pointers aren't reset automatically! Will definitely fix this.

I think it is also desirable to hard reset the radio peripheral on soft reset, similar to the other peripherals - as it might be mid-transmit or receive otherwise. Will do this the deinit function as well, though.

dpgeorge commented on ports/stm32/spi.c:
```diff
@@ -502,6 +533,14 @@ void spi_deinit(const spi_t *spi_obj) {
+        __HAL_RCC_SUBGHZSPI_RELEASE_RESET();
+        __HAL_RCC_SUBGHZSPI_CLK_DISABLE();
+        HAL_NVIC_DisableIRQ(SUBGHZSPI_IRQn);
+
```
remove unnecessary blank line

dpgeorge commented on ports/stm32/subghz.c:
```diff
@@ -0,0 +1,131 @@
+    if (callback != mp_const_none) {
+        // Passing dummy 'pin' argument of None, to keep
+        // compatibility with machine.Pin.isr() handlers.
+        mp_call_function_1(callback, mp_const_none);
```
since this is a hard IRQ, it needs further protection:
```c
mp_sched_lock();
gc_lock();
mp_call_function_1_protected(...);
gc_unlock();
mp_sched_unlock();
```

projectgus replied:
:facepalm: I switched this at the last minute, clearly without thinking! Thx.

dpgeorge commented:
I'm not sure the subghz_ext_hse32 function is needed: it feels like this should be part of the board configuration in mpconfigboard.h.  All other boards configure their system clock source in that file and it's fixed at build time.  Because when you make a new board, the clock source is pretty much fixed by the board design, and then you define a board configuration around that design.  In other words, it's highly unlikely that this HSE value needs to be set dynamically.  If someone has firmware for a WL55 board then it should already have it's clock configured for that board.

For our NUCLEO_WL55 definition, it should enable HSE using the TCXO as part of SystemClock_Config(), and probably use that instead of MSI (I was lazy in e0a0719416d6a936ade8c30314b7e9c90bfbbf23 and just went for MSI because it was easy).

Also note that the TCXO will only work if PB0 is pulled high to power it!  That should also be done as part of the board initialisation in C, eg in boards/NUCLEO_WL55/board.c.

(We could in principle have a GENERIC_WL55 firmware image that runs on all WL55 boards, and starts in MSI mode, and then python boot.py configures low-level clock sources etc.  But at the moment that's not how the stm32 port is structured.)

dpgeorge commented:
Related to the ext-hse: is this board the only one that uses the dio3 options in the micropython-lib lora driver?  If so, maybe all that related code can be removed from the python driver?

projectgus commented:
> I'm not sure the subghz_ext_hse32 function is needed: it feels like this should be part of the board configuration in mpconfigboard.h. All other boards configure their system clock source in that file and it's fixed at build time. Because when you make a new board, the clock source is pretty much fixed by the board design, and then you define a board configuration around that design
>
> For our NUCLEO_WL55 definition, it should enable HSE using the TCXO as part of SystemClock_Config(), and probably use that instead of MSI (I was lazy in https://github.com/micropython/micropython/commit/e0a0719416d6a936ade8c30314b7e9c90bfbbf23 and just went for MSI because it was easy).
>
> Also note that the TCXO will only work if PB0 is pulled high to power it! That should also be done as part of the board initialisation in C, eg in boards/NUCLEO_WL55/board.c.

OK, can do. I missed that this board is only running 4MHz for now! Just to make everyone is on the same page, my understanding is there's two ways to run this board with LoRa (without cutting or closing solder jumpers):

1. The way it is currently implemented, the MCU runs from a different clock source (i.e. MSI, HSI) and the radio controls PB0_VDDTCXO pin automatically. This probably marginally lower power consumption than running the MCU from HSE32 where the TCXO has to run all of the time.
2. MCU drives PB0 high, permanently powers the TCXO and uses this for the system clock and radio clock source.

Compared to other power consumption considerations, this is probably a pretty small difference so I don't see any problem switching at build time to 2. If it's simple to do it then I'll try and make it as easy as possible to switch in a board config for other boards, low power designs, etc.

> Related to the ext-hse: is this board the only one that uses the dio3 options in the micropython-lib lora driver? If so, maybe all that related code can be removed from the python driver?

Not exactly, some of the external SX126x modules use this feature as well so the radio can automatically reduce power consumption. I have one here that I've tested with (the Waveshare Pico SX1262 module.) It's probably technically possible to power the TCXO on this module from a microcontroller GPIO or 3V3 pin connected to DIO3 instead, same as here. So if you're worried about the code size then I could remove it and document that. On balance I think it's still useful, though.

dpgeorge commented:
> I missed that this board is only running 4MHz for now!

Ah, so did I!  You can verify the freq by running `machine.freq()`.

It would be good to at least feed the MSI into the PLL and run it faster.  But IMO much better to use the TCXO (temperature compensated oscillator) in HSE bypass mode, to feed into the PLL on the WL55 and run the CPU at its maximum of 48MHz (which is pretty low anyway, still very low power).  Then also use this clock for the radio.  (Yes that requires driving PB0 high permanently.)

> Not exactly, some of the external SX126x modules use this feature as well so the radio can automatically reduce power consumption

OK, so let's leave the dio3 feature as-is.

projectgus commented:
> >  I missed that this board is only running 4MHz for now!

> Ah, so did I! You can verify the freq by running machine.freq().

Oh, you're right - I was just coming to edit my comment and say "oh, I see the PLL is enabled!". Although I haven't confirmed this.

> IMO much better to use the TCXO (temperature compensated oscillator) in HSE bypass mode, to feed into the PLL on the WL55 and run the CPU at its maximum of 48MHz (which is pretty low anyway, still very low power). Then also use this clock for the radio. (Yes that requires driving PB0 high permanently.)

OK, will do.

projectgus committed:
stm32/powerctrlboot: Support STM32WL system clock from HSE+PLL.

Switches default on the NUCLEO_WL55 board to use the HSE oscillator powered
from PB0_VDDTCXO pin.

Build-time configuration can select from MSI internal oscillator (previous
default), HSE via crystal, or HSE bypass with TCXO powered from PB0_VDDTCXO
pin (new default)

Signed-off-by: Angus Gratton <angus@redyak.com.au>

projectgus commented:
@dpgeorge Updated, PTAL again.

Setting up PB0_VDDTCXO to behave correctly took a bit more work than expected, but verified both `MICROPY_HW_CLK_USE_HSE` and `!MICROPY_HW_CLK_USE_HSE` work correctly on Nucleo-WL55 board now, including the LoRa modem.

dpgeorge commented on ports/stm32/powerctrlboot.c:
```diff
@@ -460,13 +460,68 @@ void SystemClock_Config(void) {
+        .Mode = GPIO_MODE_ANALOG,
+        .Pull = GPIO_NOPULL,
+    };
+    HAL_GPIO_Init(GPIOB, &GPIO_InitStructure);
```
Please use `mp_hal_pin_config(pin_B0, MP_HAL_PIN_MODE_ANALOG, MP_HAL_PIN_PULL_NONE, 0)`.

projectgus replied:
I took PB0 out of `pins.csv` in this PR because (at least in the default board configuration) it's easy to change the pin mode and disable the system clock.

Do you prefer `PB0` stays accessible from Python, or is there another way to resolve this?

dpgeorge replied:
You can use `,-PB0` in `pins.csv` to make the pin accessible only from C.

dpgeorge replied:
(see the comment at the top of `ports/stm32/boards/make-pins.py` for more info)

projectgus replied:
Oh, neat! Done.

dpgeorge commented on ports/stm32/powerctrlboot.c:
```diff
@@ -460,13 +460,68 @@ void SystemClock_Config(void) {
+    }
+
+    __HAL_RCC_SYSCLK_CONFIG(RCC_SYSCLKSOURCE_PLLCLK);
+    while (__HAL_RCC_GET_SYSCLK_SOURCE() != RCC_SYSCLKSOURCE_STATUS_PLLCLK) {
```
Please use the `LL_RCC_xxx` functions instead of `__HAL_RCC_xxx` (see just below for examples).  The LL is much preferred over the old HAL.

projectgus replied:
@dpgeorge I switched from LL to HAL here because there's some register bit sequencing in the ` __HAL_RCC_HSE_CONFIG` macro that seemed unnecessary to recreate here, similarly the PLL calculation macro. However, I'll switch it back.

(As I'm sure you realise, various parts of the stm32 port contain different configuration methods so it's hard to guess which one is the preferred level to integrate against.)

projectgus replied:
Rewritten to use LL[*], confirmed system frequency and LoRa modem functions are still the same.

[*] Except PLL config which is raw register write as the only LL method seems to be inexplicably split into three overlapping register operations.

dpgeorge replied:
But it should be easy to at least just copy the code below that sets the MSI as the clock source.  That will use the same number of lines and be consistent with how MSI is done.

For the PLL... it can stay using HAL if that's simpler (but I think there's also a special LL function to configure the PLL values).

projectgus replied:
Looks like our comments crossed, but let me know if you want the PLL config lines converted back to HAL.

dpgeorge commented:
Thanks for updating, and for testing.  Now merged.

gorgonaut04 commented:
Sorry to be a bother but the timing on this couldn't be better for me, just selected this MCU for a new project. Would love to see documentation for this added to the docs website with some examples of using the radio!

dpgeorge commented:
@gorgonaut04 please see this PR here: https://github.com/micropython/micropython-lib/pull/572, and also the existing `lora` module in `micropython-lib`: https://github.com/micropython/micropython-lib/tree/master/micropython/lora.  It has docs and examples.
